pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -4014,6 +4010,31 @@ public void readEntryFailed(ManagedLedgerException exception, Object ctx) {
         return completableFuture;
     }
 
+    public synchronized CompletableFuture<Void> triggerCompactionWithCheckHasMoreMessages() {
+        return getLastDispatchablePosition().thenCombine(topicCompactionService.getLastCompactedPosition(),
```

Comment:
`checkBeforetriggerCompaction` is better ?

Reply:
It seems that it didn't specify what to check.